lib/tree.rb

```diff
@@ -2,61 +2,191 @@ class TreeNode
   attr_reader :key, :value
   attr_accessor :left, :right
 
-   def initialize(key, val)
+  def initialize(key, val)
     @key = key
     @value = val
     @left = nil
     @right = nil
-   end
+  end
 end
 
 class Tree
   attr_reader :root
+
   def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: if balanced, O(log h) where h is the height of the tree
+  # Space Complexity: if balanced, O(log h) where h is the height of the tree
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+  end
+
+  def add_helper(current, new_node)
+    if new_node.key <= current.key
+      if current.left
+        add_helper(current.left, new_node)
+      else
+        current.left = new_node
+      end
+    else
+      if current.right
+        add_helper(current.right, new_node)
+      else
+        current.right = new_node
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(h) if  balanced, where h is the height of the tree
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return nil if @root.nil?
+    current = @root
+    while !current.nil?
+      if key == current.key
+        return current.value
+      elsif key <= current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n^2) where n is the number of nodes
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    stack = []
+    output = []
+    current = root
+    while current || !stack.empty?
+      while !current.nil?
+        stack.push(current)
+        current = current.left
+      end
+      current = stack.pop
+      new_hash = Hash.new
+      new_hash[:key] = current.key
+      new_hash[:value] = current.value
+      output.push(new_hash);  #storing the data
+      current = current.right
+    end
+    return output
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: if well balanced, O(h) where h is the height of the tree
```

Comment:
You could also say O(log n) if it's balanced, except that since you're going to visit each node in the tree, it has to be O(n)